File: relay/internal/channel/fragment.go
package channel

import (
	"errors"
	"fmt"
)

// Fragment wire format: flags(1 byte) || data. Bit 0 marks the last
// fragment of a frame. Fragments of one frame are contiguous on a channel.
const (
	FlagLast     byte = 0b0000_0001
	FragmentData      = MaxNoisePayload - 1
)

var (
	ErrEmptyFragment = errors.New("fragment: empty message")
	ErrTooLarge      = errors.New("fragment: reassembled frame would exceed the limit")
)

// Fragments splits encoded frame bytes; it always yields at least one.
func Fragments(b []byte) [][]byte {
	if len(b) == 0 {
		return [][]byte{{FlagLast}}
	}
	var out [][]byte
	for len(b) > 0 {
		n := min(len(b), FragmentData)
		chunk := make([]byte, 1, n+1)
		chunk = append(chunk, b[:n]...)
		b = b[n:]
		if len(b) == 0 {
			chunk[0] = FlagLast
		}
		out = append(out, chunk)
	}
	return out
}

// Reassembler is the bounded reassembly buffer of one channel.
type Reassembler struct {
	limit  int
	buffer []byte
}

func NewReassembler(limit int) *Reassembler { return &Reassembler{limit: limit} }

// Push consumes one fragment; returns the frame bytes on the last one.
func (r *Reassembler) Push(fragment []byte) (frame []byte, done bool, err error) {
	if len(fragment) == 0 {
		return nil, false, ErrEmptyFragment
	}
	flags, data := fragment[0], fragment[1:]
	if flags&^FlagLast != 0 {
		return nil, false, fmt.Errorf("fragment: unknown flags %#04x", flags)
	}
	if len(r.buffer)+len(data) > r.limit {
		r.buffer = nil
		return nil, false, ErrTooLarge
	}
	r.buffer = append(r.buffer, data...)
	if flags&FlagLast == 0 {
		return nil, false, nil
	}
	out := r.buffer
	r.buffer = nil
	return out, true, nil
}

// MidFrame reports whether a frame is partially received.
func (r *Reassembler) MidFrame() bool { return len(r.buffer) > 0 }

File: core/crates/arveil-core/src/channel/fragment.rs
//! Fragmentation of frames into Noise payloads.
//!
//! Wire format of one fragment: `flags(1 byte) || data`. `flags` bit 0 is
//! set on the last fragment of a frame. Fragments of one frame are sent
//! back to back on the same channel; interleaving is not allowed, which
//! keeps the reassembler a single bounded buffer.

use super::noise::MAX_NOISE_PAYLOAD;

pub const FLAG_LAST: u8 = 0b0000_0001;
/// Data bytes per fragment.
pub const FRAGMENT_DATA: usize = MAX_NOISE_PAYLOAD - 1;

#[derive(Debug, thiserror::Error)]
pub enum FragmentError {
    #[error("fragment: empty message")]
    Empty,
    #[error("fragment: unknown flags {0:#04x}")]
    UnknownFlags(u8),
    #[error("fragment: reassembled frame would exceed {limit} bytes")]
    TooLarge { limit: usize },
}

/// Split encoded frame bytes into fragments. Always yields at least one.
pub fn fragments(bytes: &[u8]) -> impl Iterator<Item = Vec<u8>> + '_ {
    let chunks: Vec<&[u8]> = if bytes.is_empty() {
        vec![&[][..]]
    } else {
        bytes.chunks(FRAGMENT_DATA).collect()
    };
    let last = chunks.len() - 1;
    chunks.into_iter().enumerate().map(move |(i, chunk)| {
        let mut out = Vec::with_capacity(chunk.len() + 1);
        out.push(if i == last { FLAG_LAST } else { 0 });
        out.extend_from_slice(chunk);
        out
    })
}

/// Bounded reassembly buffer for one channel.
pub struct Reassembler {
    limit: usize,
    buffer: Vec<u8>,
}

impl Reassembler {
    pub fn new(limit: usize) -> Self {
        Self {
            limit,
            buffer: Vec::new(),
        }
    }

    /// Push one fragment. Returns the complete frame bytes on the last one.
    pub fn push(&mut self, fragment: &[u8]) -> Result<Option<Vec<u8>>, FragmentError> {
        let (&flags, data) = fragment.split_first().ok_or(FragmentError::Empty)?;
        if flags & !FLAG_LAST != 0 {
            return Err(FragmentError::UnknownFlags(flags));
        }
        if self.buffer.len() + data.len() > self.limit {
            self.buffer.clear();
            return Err(FragmentError::TooLarge { limit: self.limit });
        }
        self.buffer.extend_from_slice(data);
        if flags & FLAG_LAST != 0 {
            Ok(Some(std::mem::take(&mut self.buffer)))
        } else {
            Ok(None)
        }
    }

    /// True while a frame is partially received.
    pub fn is_mid_frame(&self) -> bool {
        !self.buffer.is_empty()
    }
}
